File: mahaam-api-java/src/main/java/mahaam/feat/plan/PlanRepo.java
package mahaam.feat.plan;

import java.sql.Timestamp;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

import org.jdbi.v3.core.result.RowView;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import mahaam.feat.plan.PlanModel.Plan;
import mahaam.feat.plan.PlanModel.PlanIn;
import mahaam.feat.plan.PlanModel.PlanType;
import mahaam.feat.task.Task;
import mahaam.feat.user.UserModel.User;
import mahaam.infra.DB;
import mahaam.infra.Log;
import mahaam.infra.Mapper;
import mahaam.infra.Req;

public interface PlanRepo {
	Plan getOne(UUID id);

	List<Plan> getMany(UUID userId, String type);

	UUID create(PlanIn plan);

	void update(PlanIn plan);

	void delete(UUID id);

	void updateDonePercent(UUID id);

	void removeFromOrder(UUID userId, UUID id);

	void updateOrder(UUID userId, String type, int oldOrder, int newOrder);

	void updateType(UUID userId, UUID id, String type);

	long getCount(UUID userId, String type);

	int updateUserId(UUID oldUserId, UUID newUserId);
}

@ApplicationScoped
class DefaultPlanRepo implements PlanRepo {

	@Inject
	DB db;

	@Override
	public UUID create(PlanIn plan) {
		String query = """
				INSERT INTO plans (id, user_id, title, starts, ends, type, status, done_percent, sort_order, created_at)
				VALUES (:id, :userId, :title, :starts, :ends, :type, :status, '0/0',
				(SELECT COUNT(1) FROM plans WHERE user_id = :userId AND type = :type), current_timestamp) """;

		UUID id = UUID.randomUUID();
		db.insert(
				query,
				Mapper.of(
						"id",
						id,
						"title",
						plan.title,
						"starts",
						toTimestamp(plan.starts),
						"ends",
						toTimestamp(plan.ends),
						"userId",
						Req.getUserId(),
						"type",
						PlanType.MAIN,
						"status",
						"Open"));

		return id;
	}

	@Override
	public void update(PlanIn plan) {
		String query = """
				UPDATE plans
				SET title = :title, starts = :starts, ends = :ends, updated_at = current_timestamp
				WHERE id = :id
				""";

		db.update(
				query,
				Mapper.of(
						"id", plan.id,
						"title", plan.title,
						"starts", toTimestamp(plan.starts),
						"ends", toTimestamp(plan.ends)));
	}

	@Override
	public Plan getOne(UUID id) {
		String query = """
				SELECT c.id c_id, c.title c_title, c.starts c_starts, c.ends c_ends, c.type c_type,
				    c.done_percent c_donePercent, c.sort_order c_sortOrder, c.user_id c_userId,
					EXISTS(SELECT 1 FROM plan_members cm WHERE cm.plan_id = c.id) AS c_isShared,
					u.id u_id, u.email u_email, u.name u_name
				FROM plans c
				LEFT JOIN users u ON c.user_id = u.id
				WHERE c.id = :id""";

		var plan = db.getJdbi().withHandle(
				handle -> {
					return handle
							.createQuery(query)
							.bind("id", id)
							.reduceRows(
									new LinkedHashMap<Plan, User>(),
									(Map<Plan, User> map, RowView rowView) -> {
										Plan c = rowView.getRow(Plan.class);
										User u = rowView.getRow(User.class);
										c.user = u;
										map.put(c, u);
										return map;
									})
							.entrySet()
							.stream()
							.map(entry -> entry.getKey())
							.findFirst()
							.orElse(null);
				});

		return plan;
	}

	@Override
	public List<Plan> getMany(UUID userId, String type) {
		try {
			String query = """
					SELECT c.id c_id, c.title c_title, c.starts c_starts, c.ends c_ends, c.type c_type,
					    c.done_percent c_donePercent, c.sort_order c_sortOrder, c.user_id c_userId,
					    EXISTS(SELECT 1 FROM plan_members cm WHERE cm.plan_id = c.id) AS c_isShared,
					    u.id u_id, u.email u_email, u.name u_name
					FROM plans c
					LEFT JOIN users u ON c.user_id = u.id
					WHERE c.user_id = :user_id AND c.type = :type
					ORDER BY c.sort_order DESC;""";

			var plans = db.getJdbi()
					.withHandle(
							handle -> {
								return handle
										.createQuery(query)
										.bind("user_id", userId)
										.bind("type", type)
										.reduceRows(
												new LinkedHashMap<Plan, User>(),
												(Map<Plan, User> map, RowView rowView) -> {
													Plan c = rowView.getRow(Plan.class);
													User u = rowView.getRow(User.class);
													c.user = u;
													map.put(c, u);
													return map;
												})
										.keySet()
										.stream()
										.collect(Collectors.toList());
							});
			return plans;
		} catch (Exception e) {
			Log.error(e.toString());
			return new ArrayList<>();
		}
	}

	@Override
	public void delete(UUID id) {
		int count = db.delete("DELETE FROM plans WHERE id = :id", Mapper.of("id", id));
		if (count > 0) {
			Log.info("Plan " + id + " deleted");
		}
	}

	@Override
	public void updateDonePercent(UUID id) {
		String query = """
				SELECT id t_id, plan_id t_planId, title t_title, done t_done, sort_order t_sortOrder,
					created_at t_createdAt, updated_at t_updatedAt FROM tasks t WHERE t.plan_id = :id""";
		List<Task> tasks = db.selectList(query, Task.class, Mapper.of("id", id));

		long done = tasks.stream().filter(task -> task.done).count();
		int notDone = tasks.size();
		String donePercent = done + "/" + notDone;
		String updateQuery = "UPDATE plans SET done_percent = :donePercent WHERE id = :id";
		db.update(updateQuery, Mapper.of("donePercent", donePercent, "id", id));
	}

	@Override
	public void removeFromOrder(UUID userId, UUID id) {
		String query = """
				UPDATE plans
				SET sort_order = sort_order - 1
				WHERE
				    user_id = :userId AND
				    type = (SELECT type FROM plans WHERE id = :id) AND
				    sort_order > (SELECT sort_order FROM plans WHERE id = :id)
				""";

		db.update(query, Mapper.of("userId", userId, "id", id));
	}

	@Override
	public void updateOrder(UUID userId, String type, int oldOrder, int newOrder) {
		String query = """
				UPDATE plans SET sort_order =
				    CASE
				        WHEN sort_order = :oldOrder THEN :newOrder
				        WHEN sort_order > :oldOrder AND sort_order <= :newOrder THEN sort_order - 1
				        WHEN sort_order >= :newOrder AND sort_order < :oldOrder THEN sort_order + 1
				        ELSE sort_order
				    END
				WHERE
				    user_id = :userId AND
				    type = :type
				""";

		db.update(
				query,
				Mapper.of(
						"userId", userId,
						"type", type,
						"oldOrder", oldOrder,
						"newOrder", newOrder));
	}

	@Override
	public void updateType(UUID userId, UUID id, String type) {
		String query = """
				UPDATE plans
				SET type = :type,
				    sort_order = (SELECT COUNT(1) FROM plans WHERE user_id = :userId AND type = :type),
				    updated_at = current_timestamp
				WHERE id = :id
				""";

		db.update(query, Mapper.of("userId", userId, "id", id, "type", type));
	}

	@Override
	public long getCount(UUID userId, String type) {
		String queryCount = "SELECT COUNT(*) FROM plans WHERE user_id = :userId AND type = :type";
		return db.selectCount(queryCount, Mapper.of("userId", userId, "type", type));
	}

	@Override
	public int updateUserId(UUID oldUserId, UUID newUserId) {
		String query = """
				UPDATE plans
				SET user_id = :newUserId,
				    sort_order = (sort_order + (SELECT count(1) FROM plans WHERE user_id = :newUserId)),
				    updated_at = current_timestamp
				WHERE user_id = :oldUserId
				""";

		return db.update(query, Mapper.of("oldUserId", oldUserId, "newUserId", newUserId));
	}

	private Timestamp toTimestamp(LocalDate dateTime) {
		if (dateTime == null) {
			return null;
		}
		return java.sql.Timestamp.valueOf(dateTime.atStartOfDay());
	}
}

File: mahaam-api-py/feat/plan/plan_repo.py
from uuid import UUID, uuid4
from typing import Protocol

from infra import db
from infra.req import req
from infra.log import Log
from feat.plan.plan_model import Plan, PlanIn, PlanType
from feat.user.user_model import User
from infra.validation import ProtocolEnforcer


class PlanRepo(Protocol):
	def select_one(self, id: UUID) -> Plan | None: ...
	def select_many(self, user_id: UUID, type: str) -> list[Plan]: ...
	def create(self, plan: PlanIn, conn=None) -> UUID: ...
	def update(self, plan: PlanIn) -> None: ...
	def delete(self, id: UUID, conn=None) -> None: ...
	def update_done_percent(self, id: UUID, conn) -> None: ...
	def remove_from_order(self, user_id: UUID, id: UUID, conn=None) -> None: ...
	def update_order(self, user_id: UUID, type: str,
					 old_order: int, new_order: int, conn=None) -> None: ...
	def update_type(self, user_id: UUID, id: UUID, type: str, conn=None) -> None: ...
	def get_count(self, user_id: UUID, type: str, conn=None) -> int: ...
	def update_user_id(self, old_user_id: UUID, new_user_id: UUID, conn=None) -> int: ...


class DefaultPlanRepo(metaclass=ProtocolEnforcer, protocol=PlanRepo):
	def create(self, plan: PlanIn, conn=None) -> UUID:
		query = """INSERT INTO plans (id, user_id, title, starts, ends, type, status, done_percent, sort_order, created_at)
			VALUES (:id, :user_id, :title, :starts, :ends, :type, :status, '0/0', 
			(SELECT COUNT(1) FROM plans WHERE user_id = :user_id AND type = :type), 
			current_timestamp)"""
		id = uuid4()
		db.DB.insert(query, {
			"id": str(id),
			"title": plan.title,
			"starts": plan.starts,
			"ends": plan.ends,
			"user_id": req.user_id,
			"type": PlanType.MAIN,
			"status": "Open"
		}, conn)
		return id

	def update(self, plan: PlanIn) -> None:
		query = "UPDATE plans SET title = :title, starts = :starts, ends = :ends, updated_at = current_timestamp WHERE id = :id"
		db.DB.update(query, {"id": str(plan.id), "title": plan.title, "starts": plan.starts, "ends": plan.ends})

	def select_one(self, id: UUID) -> Plan | None:
		query = """
			SELECT c.id, c.title, c.starts, c.ends, c.type, c.done_percent, c.sort_order,
				EXISTS(SELECT 1 FROM plan_members cm WHERE cm.plan_id = c.id) AS is_shared,
				u.id user_id, u.email user_email, u.name user_name
			FROM plans c
			LEFT JOIN users u ON c.user_id = u.id
			WHERE c.id = :id"""

		coll = db.DB.select_one(dict, query, {"id": str(id)})
		if coll is None:
			return None
		
		plan = Plan(
			id=coll["id"],
			title=coll["title"],
			starts=coll["starts"],
			ends=coll["ends"],
			type=coll["type"],
			done_percent=coll["done_percent"],
			sort_order=coll["sort_order"],
			is_shared=coll["is_shared"],
			user=User(id=coll["user_id"], email=coll["user_email"], name=coll["user_name"])
		)
		return plan


	def select_many(self, user_id: UUID, type: str) -> list[Plan]:
		try:
			query = """
			SELECT c.id, c.title, c.starts, c.ends, c.type, c.done_percent, c.sort_order,
				EXISTS(SELECT 1 FROM plan_members cm WHERE cm.plan_id = c.id) AS is_shared,
				u.id user_id, u.email user_email, u.name user_name
			FROM plans c
			LEFT JOIN users u ON c.user_id = u.id
			WHERE c.user_id = :user_id AND c.type = :type
			ORDER BY c.sort_order DESC"""

			dic = db.DB.select_many(dict, query, {"user_id": str(user_id), "type": type})
			plans = []
			for row in dic:
				plan = Plan(
					id=row["id"],
					title=row["title"],
					starts=row["starts"],
					ends=row["ends"],
					type=row["type"],
					done_percent=row["done_percent"],
					sort_order=row["sort_order"],
					is_shared=row["is_shared"],
					user=User(id=row["user_id"], email=row["user_email"],name=row["user_name"])
				)
				plans.append(plan)
			return plans
		except Exception as e:
			Log.error(str(e))
			return []

	def delete(self, id: UUID, conn=None) -> None:
		count = db.DB.delete(
			"DELETE FROM plans WHERE id = :id", {"id": str(id)}, conn)
		if count > 0:
			Log.info(f"Plan {id} deleted")

	def update_done_percent(self, id: UUID, conn) -> None:
		query = "SELECT * FROM tasks WHERE plan_id = :id"
		tasks = db.DB.select_many(dict, query, {"id": str(id)}, conn)

		done = sum(1 for task in tasks if task["done"])
		not_done = len(tasks)
		done_percent = f"{done}/{not_done}"
		update_query = "UPDATE plans SET done_percent = :donePercent WHERE id = :id"
		db.DB.update(update_query, {"donePercent": done_percent, "id": str(id)}, conn)

	def remove_from_order(self, user_id: UUID, id: UUID, conn=None) -> None:
		query = """
			UPDATE plans SET sort_order = sort_order - 1 
			WHERE user_id = :user_id AND type = (SELECT type FROM Plans WHERE id =:id) 
				AND sort_order > (SELECT sort_order FROM plans WHERE id =:id)"""
		db.DB.update(query, {"user_id": str(user_id), "id": str(id)}, conn)

	def update_order(self, user_id: UUID, type: str, old_order: int, new_order: int, conn=None) -> None:
		query = """
			UPDATE plans SET sort_order = 
				CASE 
					WHEN sort_order = :old_order THEN :new_order
					WHEN sort_order > :old_order AND sort_order <= :new_order THEN sort_order - 1
					WHEN sort_order >= :new_order AND sort_order < :old_order THEN sort_order + 1
					ELSE sort_order
				END
			WHERE 
				user_id = :user_id AND 
				type = :type"""
		db.DB.update(query, {
			"user_id": str(user_id), 
			"type": type, 
			"old_order": old_order, 
			"new_order": new_order
		}, conn)

	def update_type(self, user_id: UUID, id: UUID, type: str, conn=None) -> None:
		query = """
			UPDATE plans SET type = :type, 
			sort_order = (SELECT COUNT(1) FROM plans WHERE user_id = :user_id AND type = :type),
			updated_at = current_timestamp WHERE id = :id"""
		db.DB.update(query, {"user_id": str(user_id), "id": str(id), "type": type}, conn)

	def get_count(self, user_id: UUID, type: str, conn=None) -> int:
		query_count = "SELECT COUNT(*) FROM plans WHERE user_id = :user_id and type = :type"
		result = db.DB.select_one(dict, query_count, {"user_id": str(user_id), "type": type}, conn)
		
		if result is None:
			return 0
		else:
			return next(iter(result.values()))

	def update_user_id(self, old_user_id: UUID, new_user_id: UUID, conn=None) -> int:
		query = """
		 	UPDATE plans SET user_id = :new_user_id,
			sort_order = (sort_order + (Select count(1) from plans where user_id=:new_user_id)),
			updated_at = current_timestamp 
			WHERE user_id = :old_user_id"""
		return db.DB.update(query, {"old_user_id": str(old_user_id), "new_user_id": str(new_user_id)}, conn)
